sky/shell/mac/tracing_controller.mm

```diff
@@ -2,119 +2,130 @@
-    controller = [[TracingController alloc] init];
-  });
-  return controller;
+  StopBaseTracing();
 }
 
-- (void)startTracing {
-  // Start Tracing
-  NSLog(@"Staring Trace");
+void TracingController::OnDataAvailable(const void* data, size_t size) {
+  if (_trace_file == nullptr) {
+    return;
+  }
 
-  base::trace_event::TraceLog::GetInstance()->SetEnabled(
-      base::trace_event::TraceConfig("*", base::trace_event::RECORD_UNTIL_FULL),
-      base::trace_event::TraceLog::RECORDING_MODE);
+  _trace_file->WriteAtCurrentPos(reinterpret_cast<const char*>(data), size);
 }
 
-- (void)stopTracing {
-  // Stop Tracing
-  NSLog(@"Stopping Trace");
-  base::trace_event::TraceLog::GetInstance()->SetDisabled();
-
-  // Save Trace File
-  NSArray* paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory,
-                                                       NSUserDomainMask, YES);
-  if (paths.count == 0) {
-    NSLog(@"Error: Could not find documents directory to write the trace file "
-          @"to");
-    return;
+void TracingController::OnDataComplete() {
+  _trace_file = nullptr;
+  _drainer = nullptr;
+}
+
+void TracingController::StartDartTracing() {
+  if (auto view = _view.lock()) {
+    view->StartDartTracing();
   }
+}
 
-  // Prepare a file path that looks sane in the documents directory
-  NSURL* pathURL = [NSURL URLWithString:paths[0]];
-  NSDateFormatter* formatter = [[NSDateFormatter alloc] init];
-  [formatter setDateFormat:@"hh_mm_s"];
-  NSString* dateString =
-      [NSString stringWithFormat:@"Trace_%@.json",
-                                 [formatter stringFromDate:[NSDate date]]];
-  [formatter release];
-  NSURL* fileURL = [NSURL URLWithString:dateString relativeToURL:pathURL];
-  NSError* error = nil;
-
-  // Create the file in the documents directory
-  BOOL created =
-      [[NSFileManager defaultManager] createFileAtPath:fileURL.absoluteString
-                                              contents:nil
-                                            attributes:nil];
-  if (!created) {
-    NSLog(@"Error: Could not create file for writing trace file to");
-    return;
+void TracingController::StopDartTracing() {
+  mojo::ScopedDataPipeProducerHandle producer;
+  mojo::ScopedDataPipeConsumerHandle consumer;
+
+  auto result = mojo::CreateDataPipe(nullptr, &producer, &consumer);
+  if (result != MOJO_RESULT_OK) {
+    DLOG(WARNING) << "Could not create data pipe for writing trace file";
   }
 
-  // Fetch a write handle to the created file
-  NSFileHandle* handle =
-      [NSFileHandle fileHandleForWritingToURL:fileURL error:&error];
-  if (error != nil) {
-    NSLog(@"Error: Could not write trace file to documents directory: %@",
-          error.localizedDescription);
-    return;
+  _drainer = std::unique_ptr<mojo::common::DataPipeDrainer>(
+      new mojo::common::DataPipeDrainer(this, std::move(consumer)));
+
+  if (auto view = _view.lock()) {
+    view->StopDartTracing(std::move(producer));
```

Comment:
Why are you using a data pipe?  Both the producer and the consumer on are on the same thread.  You can just pass in a function pointer.